File: katas/csharp/OOP/polymorphism/polymorphism-good.cs
using System;
using System.Collections.Generic;

// Cumplimiento de Polimorfismo: Comportamiento específico sin condicionales
// ✅ Solución: Cada clase implementa su propio comportamiento

// ✅ Clase base define la interfaz común
public abstract class Animal
{
  protected string name;

  public Animal(string name)
  {
    this.name = name;
  }

  // ✅ Métodos abstractos - cada hijo DEBE implementarlos
  public abstract void MakeSound();
  public abstract void Feed();
  public abstract void Move();

  // ✅ Método común para todos
  public void Introduce()
  {
    Console.WriteLine($"\nSoy {name}:");
  }
}

// ✅ Cada clase implementa su comportamiento específico
public class Dog : Animal
{
  public Dog(string name) : base(name)
  {
  }

  public override void MakeSound()
  {
    Console.WriteLine($"{name} dice: ¡Guau guau!");
  }

  public override void Feed()
  {
    Console.WriteLine($"{name} está comiendo croquetas");
  }

  public override void Move()
  {
    Console.WriteLine($"{name} está corriendo");
  }
}

public class Cat : Animal
{
  public Cat(string name) : base(name)
  {
  }

  public override void MakeSound()
  {
    Console.WriteLine($"{name} dice: ¡Miau miau!");
  }

  public override void Feed()
  {
    Console.WriteLine($"{name} está comiendo pescado");
  }

  public override void Move()
  {
    Console.WriteLine($"{name} está saltando");
  }
}

public class Bird : Animal
{
  public Bird(string name) : base(name)
  {
  }

  public override void MakeSound()
  {
    Console.WriteLine($"{name} dice: ¡Pío pío!");
  }

  public override void Feed()
  {
    Console.WriteLine($"{name} está comiendo semillas");
  }

  public override void Move()
  {
    Console.WriteLine($"{name} está volando");
  }
}

// ✅ Procesador sin condicionales - usa polimorfismo
public class AnimalProcessor
{
  // ✅ Método genérico - funciona con cualquier Animal
  public void ProcessAnimals(List<Animal> animals)
  {
    foreach (var animal in animals)
    {
      animal.Introduce();
      animal.MakeSound(); // ✅ Llama al método correcto automáticamente
      animal.Feed(); // ✅ Sin if/else
      animal.Move(); // ✅ Sin switch
    }
  }
}

// ✅ Uso limpio sin condicionales
class Program
{
  static void Main()
  {
    Console.WriteLine("=== Cumplimiento de Polimorfismo ===");

    var dog = new Dog("Rex");
    var cat = new Cat("Luna");
    var bird = new Bird("Piolín");

    // ✅ Array de tipo Animal - polimorfismo en acción
    var animals = new List<Animal> { dog, cat, bird };

    var processor = new AnimalProcessor();

    // ✅ Un solo método procesa todos los tipos sin verificar
    processor.ProcessAnimals(animals);

    // ✅ Agregar nuevo animal es fácil - solo crear la clase
    var fish = new Fish("Nemo");
    Console.WriteLine("\n=== Nuevo animal agregado sin modificar código existente ===");

    // ✅ Funciona inmediatamente sin cambiar AnimalProcessor
    processor.ProcessAnimals(new List<Animal> { fish });

    // ✅ Beneficios:
    // - Sin if/else ni switch
    // - Agregar nuevos animales no modifica código existente
    // - Cada clase tiene su lógica encapsulada
    // - Respeta Open/Closed Principle
    // - Fácil de mantener y escalar
  }
}

public class Fish : Animal
{
  public Fish(string name) : base(name)
  {
  }

  public override void MakeSound()
  {
    Console.WriteLine($"{name} hace burbujas: glu glu");
  }

  public override void Feed()
  {
    Console.WriteLine($"{name} está comiendo algas");
  }

  public override void Move()
  {
    Console.WriteLine($"{name} está nadando");
  }
}

File: katas/typescript/src/OOP/polymorphism/polymorphism-good.ts
// Cumplimiento de Polimorfismo: Comportamiento específico sin condicionales
// ✅ Solución: Cada clase implementa su propio comportamiento

// ✅ Clase base define la interfaz común
abstract class Animal {
  protected name: string;

  constructor(name: string) {
    this.name = name;
  }

  // ✅ Métodos abstractos - cada hijo DEBE implementarlos
  abstract makeSound(): void;
  abstract feed(): void;
  abstract move(): void;

  // ✅ Método común para todos
  public introduce(): void {
    console.log(`\nSoy ${this.name}:`);
  }
}

// ✅ Cada clase implementa su comportamiento específico
class Dog extends Animal {
  public makeSound(): void {
    console.log(`${this.name} dice: ¡Guau guau!`);
  }

  public feed(): void {
    console.log(`${this.name} está comiendo croquetas`);
  }

  public move(): void {
    console.log(`${this.name} está corriendo`);
  }
}

class Cat extends Animal {
  public makeSound(): void {
    console.log(`${this.name} dice: ¡Miau miau!`);
  }

  public feed(): void {
    console.log(`${this.name} está comiendo pescado`);
  }

  public move(): void {
    console.log(`${this.name} está saltando`);
  }
}

class Bird extends Animal {
  public makeSound(): void {
    console.log(`${this.name} dice: ¡Pío pío!`);
  }

  public feed(): void {
    console.log(`${this.name} está comiendo semillas`);
  }

  public move(): void {
    console.log(`${this.name} está volando`);
  }
}

// ✅ Procesador sin condicionales - usa polimorfismo
class AnimalProcessor {
  // ✅ Método genérico - funciona con cualquier Animal
  public processAnimals(animals: Animal[]): void {
    animals.forEach((animal) => {
      animal.introduce();
      animal.makeSound(); // ✅ Llama al método correcto automáticamente
      animal.feed(); // ✅ Sin if/else
      animal.move(); // ✅ Sin switch
    });
  }
}

// ✅ Uso limpio sin condicionales
console.log("=== Cumplimiento de Polimorfismo ===");

const dog = new Dog("Rex");
const cat = new Cat("Luna");
const bird = new Bird("Piolín");

// ✅ Array de tipo Animal - polimorfismo en acción
const animals: Animal[] = [dog, cat, bird];

const processor = new AnimalProcessor();

// ✅ Un solo método procesa todos los tipos sin verificar
processor.processAnimals(animals);

// ✅ Agregar nuevo animal es fácil - solo crear la clase
class Fish extends Animal {
  public makeSound(): void {
    console.log(`${this.name} hace burbujas: glu glu`);
  }

  public feed(): void {
    console.log(`${this.name} está comiendo algas`);
  }

  public move(): void {
    console.log(`${this.name} está nadando`);
  }
}

const fish = new Fish("Nemo");
console.log("\n=== Nuevo animal agregado sin modificar código existente ===");

// ✅ Funciona inmediatamente sin cambiar AnimalProcessor
processor.processAnimals([fish]);

// ✅ Beneficios:
// - Sin if/else ni switch
// - Agregar nuevos animales no modifica código existente
// - Cada clase tiene su lógica encapsulada
// - Respeta Open/Closed Principle
// - Fácil de mantener y escalar

export { Animal, Dog, Cat, Bird, Fish, AnimalProcessor };
